quadratic/quadratic_problem.py

```diff
@@ -65,6 +67,87 @@ def from_params(cls, G: np.ndarray, c: np.ndarray, A: np.ndarray,
-    def is_inequality_constrained(self):
+    def is_inequality_constrained(self) -> bool:
         """Return whether or not any of the imposed constraints are inequalities."""
-        return any(constraint.equation_type is not EquationType.EQ for constraint in self.constraints)
+        return any(constraint.equation_type is not EquationType.EQ
+                   for constraint in self.constraints)
+
+    def find_x0(self, initial_guess: np.ndarray = None) -> np.ndarray:
+        """Find initial solution
+
+        Args:
+            initial_guess: Point to start from.
+
+        Returns:
+            np.ndarray - Feasible point.
+        """
+        if self.x0 is not None:
+            return self.x0
+        if initial_guess is None:
+            return find_x0(self, standardized=False)
+
+        z = np.array([self.compute_z_i(x=initial_guess, constraint=constraint)
+                      for constraint in self.constraints])
+
+        e = np.concatenate([np.zeros(len(initial_guess)), np.ones(len(z))])
+
+        gamma = np.array([self.compute_gamma_i(x=initial_guess, constraint=constraint)
+                          for constraint in self.constraints])
+
+        # We bring this into a sort of standard form by having the gammas
+        # being unit-vectors, quasi.
+        constraints = [LinearConstraint(c=LinearCallable(
+            a=np.concatenate([constraint.c.a, unit_vector * gamma_i]),
+            b=constraint.c.b),
+            equation_type=constraint.equation_type
+        )
+            for constraint, unit_vector, gamma_i
+            in zip(self.constraints, np.eye(len(self.constraints)), gamma)]
+
+        x0 = np.concatenate([initial_guess, z])
+
+        identity = np.eye(len(x0))
+
+        # Add positivity constraints for z.
+        constraints += [LinearConstraint(c=LinearCallable(
+            a=unit_vector,
+            b=0),
+            equation_type=EquationType.GE)
+            for unit_vector in identity[len(initial_guess):]]
+
+        sub_problem = LinearProblem(constraints=constraints,
+                                    x0=x0,
+                                    solution=None,
+                                    c=e, n=len(x0))
+
+        solution, _ = minimize_linear_problem(sub_problem, standardized=False)
+
+        return solution[:len(initial_guess)]
+
+    @staticmethod
+    def compute_gamma_i(x: np.ndarray, constraint: LinearConstraint) -> float:
+        """Compute entry of gamma for the warm start method on page 473.
+
+        Args:
+            x: Point, usually being the initial guess.
+            constraint: Some constraint.
+        """
+        a, b = constraint.c.a, constraint.c.b
+        if constraint.equation_type == EquationType.EQ:
+            return -np.sign(np.inner(a, x) - b)
+        else:
+            return 1
+
+    @staticmethod
+    def compute_z_i(x: np.ndarray, constraint: LinearConstraint) -> float:
+        """Compute entry of z for the warm start method on page 473.
```

Comment:
(minor) ned unbedingt "warm start" - des is mehr a SQP ding, was zwar mit QP interagiert, aber aus QP sicht gibts glaub i ka warm start